src/common/java/net/minecraftforge/gradle/common/task/DownloadAssets.java

```diff
@@ -48,21 +49,32 @@ public void run() throws IOException, InterruptedException {
         AssetIndex index = Utils.loadJson(getIndex(), AssetIndex.class);
         List<String> keys = new ArrayList<>(index.objects.keySet());
         Collections.sort(keys);
-        ExecutorService executorService = Executors.newFixedThreadPool(16);
-        CopyOnWriteArrayList<String> downloadingFailedURL = new CopyOnWriteArrayList<>();
+        File assetsPath = new File(Utils.getMCDir(), "/assets/objects");
+        ExecutorService executorService = Executors.newFixedThreadPool(8);
+        CopyOnWriteArrayList<String> failedDownloads = new CopyOnWriteArrayList<>();
         for (String key : keys) {
             Asset asset = index.objects.get(key);
             File target = Utils.getCache(getProject(), "assets", "objects", asset.getPath());
-            if (!target.exists()) {
+            if (!target.exists() || !HashFunction.SHA1.hash(target).equals(asset.hash)) {
                 URL url = new URL(RESOURCE_REPO + asset.getPath());
                 Runnable copyURLtoFile = () -> {
                     try {
-                        getProject().getLogger().lifecycle("Downloading: " + url + " Asset: " + key);
-                        FileUtils.copyURLToFile(url, target, 10_000, 5_000);
-
+                        File localFile = FileUtils.getFile(assetsPath + File.separator + asset.getPath());
+                        if (localFile.exists()) {
+                            getProject().getLogger().lifecycle("Copying local object: " + asset.getPath() + " Asset: " + key);
+                            FileUtils.copyFile(localFile, target);
+                        } else {
+                            getProject().getLogger().lifecycle("Downloading: " + url + " Asset: " + key);
+                            FileUtils.copyURLToFile(url, target, 10_000, 5_000);
```

Comment:
I'm considering that we should use a unified timeout value instead of hardcode.
#705